### PR TITLE
[go_router_builder] Adds support for required $extra parameter

### PR DESCRIPTION
**Pull request by dancamdev**
This PR includes support for having type-safe non-nullable `$extra` parameters when working with `TypedGoRoutes`. Optional `$extra` is still supported so this isn't a breaking change.

```Dart
@TypedGoRoute<RequiredExtraRoute>(path: '/requiredExtra')
class RequiredExtraRoute extends GoRouteData {
  const RequiredExtraRoute({required this.$extra});

  final Extra $extra;

  @override
  Widget build(BuildContext context, GoRouterState state) =>
      RequiredExtraScreen(extra: $extra);
}
```

*List which issues are fixed by this PR. You must list at least one issue.*
[#117261](https://github.com/flutter/flutter/issues/117261)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by chunhtai** [comment]
Hi @hangyujin Can you take a look at this PR?

**Review by hangyujin — APPROVED** [review]
LGTM

**Comment by dancamdev** [comment]
Fixed the merge conflicts with `1.2.0`

File: packages/go_router_builder/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 1.2.1
+
+* Supports opt-in required extra parameters. [#117261]
```
Comment (chunhtai) [inline packages/go_router_builder/CHANGELOG.md:3]: You would need to provide a link as well

File: packages/go_router_builder/example/lib/extra_example.dart [hunk 2]
```diff
@@ -0,0 +1,123 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// ignore_for_file: public_member_api_docs, always_specify_types
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/extra_example.dart:5]: Remove always_specify_types ?

**Comment by dancamdev** [comment]
Thanks @Michele-x98 for the fix!

**Review by chunhtai — APPROVED** [review]
LGTM, once the ci failure are fixed

**Comment by dancamdev** [comment]
@chunhtai should be good to go

**Comment by chunhtai** [comment]
It looks like there are still some failures

**Comment by dancamdev** [comment]
> It looks like there are still some failures

Due to the following:

> info - lib/src/route_config.dart:99:57 - 'element' is deprecated and shouldn't be used. Use element2 instead. Try replacing the use of the deprecated member with the replacement. - deprecated_member_use

But this is a CI issue because `element2` is deprecated and `element` should be used instead - the go_router_builder codebase is using `element` across the board.

File: packages/go_router_builder/lib/src/route_config.dart [hunk 3]
```diff
@@ -96,10 +96,6 @@ class RouteConfig {
       );
     }
 
-    // TODO(kevmoo): validate that this MUST be a subtype of `GoRouteData`
-    // TODO(stuartmorgan): Remove this ignore once 'analyze' can be set to
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:100]: I think this ignore and todo should stay in order to pass the analyzer

File: packages/go_router_builder/lib/src/route_config.dart [hunk 4]
```diff
@@ -474,16 +467,10 @@ GoRouteData.\$route(
...
-  // TODO(stuartmorgan): Remove this ignore once 'analyze' can be set to
-  // 5.2+ (when Flutter 3.4+ is on stable).
-  // ignore: deprecated_member_use
   final String enumName = type.element.name;
 
   final StringBuffer buffer = StringBuffer('const ${enumMapName(type)} = {');
 
-  // TODO(stuartmorgan): Remove this ignore once 'analyze' can be set to
-  // 5.2+ (when Flutter 3.4+ is on stable).
-  // ignore: deprecated_member_use
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:486]: same

File: packages/go_router_builder/lib/src/type_helpers.dart [hunk 5]
```diff
@@ -86,19 +86,16 @@ String encodeField(PropertyAccessorElement element) {
   );
 }
 
-/// Gets the name of the `const` map generated to help encode [Enum] types.
-// TODO(stuartmorgan): Remove this ignore once 'analyze' can be set to
-// 5.2+ (when Flutter 3.4+ is on stable).
-// ignore: deprecated_member_use
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/type_helpers.dart:92]: same

**Comment by dancamdev** [comment]
@chunhtai Got it working!

**Pull request merged**